File: dashlane-project/Dashlane/src/main/java/com/dashlane/ui/dialogs/fragment/WaiterDialogFragment.java
package com.dashlane.ui.dialogs.fragment;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.dashlane.R;
import com.dashlane.dagger.singleton.SingletonProvider;

import java.lang.ref.WeakReference;

import androidx.annotation.NonNull;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.FragmentManager;

public class WaiterDialogFragment extends DialogFragment {

    public static final String TAG = WaiterDialogFragment.class.getName();

    private static final String ARGS_TITLE = "ARGS_TITLE";
    private static final String ARGS_DESCRIPTION = "ARGS_DESCRIPTION";

    @NonNull
    private static WeakReference<WaiterDialogFragment> sLastDialog = new WeakReference<>(null);

    private String mQuestion;
    private View mView;

    private static WaiterDialogFragment newInstance(String title, String question) {
        WaiterDialogFragment fragment = new WaiterDialogFragment();
        Bundle args = new Bundle();
        args.putString(ARGS_TITLE, title);
        args.putString(ARGS_DESCRIPTION, question);
        fragment.setArguments(args);
        return fragment;
    }

    public static void showWaiter(boolean cancelable, String title, String question,
                                  FragmentManager fm) {
        WaiterDialogFragment dialog = (WaiterDialogFragment) fm.findFragmentByTag(WaiterDialogFragment.TAG);
        if (dialog == null) {
            dialog = WaiterDialogFragment.newInstance(title, question);
            dialog.setCancelable(cancelable);
            dialog.show(fm, WaiterDialogFragment.TAG);
            sLastDialog = new WeakReference<>(dialog);
        }
    }

    public static void dismissWaiter(FragmentManager fm) {
        if (fm == null) {
            return;
        }
        WaiterDialogFragment dialog = (WaiterDialogFragment) fm.findFragmentByTag(WaiterDialogFragment.TAG);
        if (dialog == null) {
            WaiterDialogFragment dialogFragment = sLastDialog.get();
            if (dialogFragment != null) {
                dialogFragment.dismiss();
            }
        } else {
            dialog.dismiss();
        }
        sLastDialog = new WeakReference<>(null);
    }

    public static void updateWaiterDescription(FragmentManager fm, boolean cancelable, String message) {
        WaiterDialogFragment dialog = (WaiterDialogFragment) fm.findFragmentByTag(WaiterDialogFragment.TAG);
        if (dialog != null) {
            dialog.setCancelable(cancelable);
            dialog.updateWaiterDescription(message);
        }
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setStyle(STYLE_NO_TITLE, 0);
        parseArguments();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        mView = ((LayoutInflater) getActivity().getSystemService(Context.LAYOUT_INFLATER_SERVICE)).inflate(R.layout
                                                                                                                   .progress_dialog,
                                                                                                           null);
        ((TextView) mView.findViewById(R.id.question)).setText(mQuestion);
        return mView;
    }

    @Override
    public void onResume() {
        super.onResume();
        
        SingletonProvider.getAnnouncementCenter().disable();
    }

    @Override
    public void onStop() {
        super.onStop();
        
        SingletonProvider.getAnnouncementCenter().restorePreviousStateIfDisabled();
    }

    public void updateWaiterDescription(String message) {
        if (mView != null) {
            ((TextView) mView.findViewById(R.id.question)).setText(message);
        }
    }

    @Override
    public void onDestroyView() {
        if (getDialog() != null)
            getDialog().setDismissMessage(null);
        super.onDestroyView();
    }

    private void parseArguments() {
        if (getArguments() != null) {
            mQuestion = getArguments().getString(ARGS_DESCRIPTION);
        }
    }

}

File: dashlane-project/Dashlane/src/main/java/com/dashlane/ui/dialogs/fragment/WaiterDialogFragment.kt
package com.dashlane.ui.dialogs.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.FragmentManager
import com.dashlane.announcements.AnnouncementCenter
import com.dashlane.databinding.ProgressDialogBinding
import dagger.hilt.android.AndroidEntryPoint
import java.lang.ref.WeakReference
import javax.inject.Inject

@AndroidEntryPoint
class WaiterDialogFragment : DialogFragment() {

    @Inject
    lateinit var announcementCenter: AnnouncementCenter

    private lateinit var progressDialogBinding: ProgressDialogBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setStyle(STYLE_NO_TITLE, 0)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        progressDialogBinding = ProgressDialogBinding.inflate(inflater, container, false)
        progressDialogBinding.question.text = arguments?.getString(ARGS_DESCRIPTION)
        return progressDialogBinding.root
    }

    override fun onResume() {
        super.onResume()
        
        announcementCenter.disable()
    }

    override fun onStop() {
        super.onStop()
        
        announcementCenter.restorePreviousStateIfDisabled()
    }

    fun updateWaiterDescription(message: String?) {
        progressDialogBinding.question.text = message
    }

    override fun onDestroyView() {
        dialog?.setDismissMessage(null)
        super.onDestroyView()
    }

    companion object {
        private var lastDialog = WeakReference<WaiterDialogFragment>(null)
        private val TAG: String = WaiterDialogFragment::class.java.name
        private const val ARGS_TITLE = "ARGS_TITLE"
        private const val ARGS_DESCRIPTION = "ARGS_DESCRIPTION"

        private fun newInstance(title: String, question: String): WaiterDialogFragment {
            val fragment = WaiterDialogFragment()
            val args = Bundle()
            args.putString(ARGS_TITLE, title)
            args.putString(ARGS_DESCRIPTION, question)
            fragment.arguments = args
            return fragment
        }

        fun showWaiter(cancelable: Boolean, title: String, question: String, fm: FragmentManager) {
            var dialog = fm.findFragmentByTag(TAG) as WaiterDialogFragment?
            if (dialog == null) {
                dialog = newInstance(title, question)
                dialog.isCancelable = cancelable
                dialog.show(fm, TAG)
                lastDialog = WeakReference<WaiterDialogFragment>(dialog)
            }
        }

        fun dismissWaiter(fm: FragmentManager) {
            val dialog = fm.findFragmentByTag(TAG) as WaiterDialogFragment?
            if (dialog != null) {
                dialog.dismiss()
            } else {
                lastDialog.get()?.dismiss()
            }
            lastDialog = WeakReference(null)
        }

        fun updateWaiterDescription(fm: FragmentManager, cancelable: Boolean, message: String?) {
            val dialog = fm.findFragmentByTag(TAG) as WaiterDialogFragment?
            if (dialog != null) {
                dialog.isCancelable = cancelable
                dialog.updateWaiterDescription(message)
            }
        }
    }
}